src/common/status.h

```diff
@@ -285,3 +282,10 @@ struct StatusOr {
   template <typename>
   friend struct StatusOr;
 };
+
+#define GET_OR_RET(...)                                         \
+  ({                                                            \
+    auto&& status = (__VA_ARGS__);                              \
+    if (!status) return std::forward<decltype(status)>(status); \
+    std::forward<decltype(status)>(status);                     \
+  }).GetValue()
```

Comment:
According to https://gcc.gnu.org/onlinedocs/gcc/Statement-Exprs.html:

> In G++, the result value of a statement expression undergoes array and function pointer decay, and is returned by value to the enclosing expression. For instance, if A is a class, then constructs a temporary A object to hold the result of the statement expression, and that is used to invoke Foo. Therefore the this pointer observed by Foo is not the address of a.

Besides, comparing:

```c++
const auto &cref = function(xxx);
const auto &cref = GET_OR_RET(xxx);
```

The latter is possible to cause `cref` a dangling reference.

If we inline this trick, said for:

```c++
ttl_ = GET_OR_RET(parser.TakeInt<int>(TTL_RANGE<int>));
```

to

```c++
res = parser.TakeInt<int>(TTL_RANGE<int>);
if (!res) return res;
ttl_ = res.GetValue();
```

It won't increase too much code.

Reply:
> The latter is possible to cause cref a dangling reference.

It is actually not a problem of this macro.

It is a problem for all C++ program, you need to care about lifetime.

Reply:
> This seems like an uncommon code snippet to me. How can I read it?

I cannot see anything "uncommon".

Reply:
You convince me. My final comment on this stuff is whether we can name it `STATUS_UNWRAP` or simply `UNWRAP`.

Reply:
I think `unwrap` in rust means that, if it has a value, return it, otherwise just abort. But here we will not abort the program, just return the error status to the upper layer. So I think it is not appropriate.